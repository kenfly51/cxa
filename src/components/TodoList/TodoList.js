import React, { useState } from "react";
import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import "./TodoList.scss";

let currentId = 0;
export const TodoList = () => {
  const [todos, updateTodoList] = useState([]);
  const [searchTerm, updateSearchTerm] = useState("");

  const renderTodoItem = todo => {
    const { name, status, id } = todo;
    return (
      <div className={`todo-item ${status ? "completed" : ""}`} key={id}>
        <div className="name">{name}</div>
        <div className="actions">
          <Button variant="danger" onClick={() => deleteTodo(todo)}>
            Delete
          </Button>
          <Button variant="success" onClick={() => completeTodo(todo)}>
            Complete
          </Button>
        </div>
      </div>
    );
  };

  const renderTodoList = () => {
    return getFilteredTodo().map(renderTodoItem);
  };

  const addNewTodo = e => {
    const todo = e.target[0].value || "";
    e.preventDefault();
    if (todo) {
      todos.push({
        id: currentId++,
        name: todo.trim(),
        status: false
      });
      e.target[0].value = "";
      updateTodoList([...todos]);
    }
  };

  const deleteTodo = ({ id }) => {
    const newTodos = todos.filter(t => {
      return t.id !== id;
    });

    updateTodoList([...newTodos]);
  };

  const completeTodo = ({ id, status }) => {
    if (!status) {
      todos.forEach(t => {
        if (t.id === id) {
          t.status = true;
        }
      });

      updateTodoList([...todos]);
    }
  };

  const renderItemList = () => {
    return <div className="empty-list">I'm lonely now</div>;
  };

  const searchTodo = e => {
    e.preventDefault();
    const searchTerm = (e.target[0].value || "").toLowerCase().trim();
    updateSearchTerm(searchTerm);
  };

  const getFilteredTodo = () => {
    return todos.filter(t => {
      return searchTerm === "" || t.name.toLowerCase().includes(searchTerm);
    });
  };

  const renderSearchbar = () => {
    return (
      <Form onSubmit={searchTodo}>
        <Form.Control placeholder="search todo" />
      </Form>
    );
  };

  return (
    <div className="todo-list">
      <h3>To Do List</h3>
      <div className="cotainer">
        <div className="search-bar">{renderSearchbar()}</div>
        {todos.length > 0 ? renderTodoList() : renderItemList()}
        <Form onSubmit={addNewTodo}>
          <Form.Control placeholder="Enter new todo item..." />
        </Form>
      </div>
    </div>
  );
};
