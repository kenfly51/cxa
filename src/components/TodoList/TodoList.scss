.todo-list {
  display: flex;
  flex-direction: column;
  .cotainer {
    display: flex;
    flex-direction: column;
    padding: 20px;

    .empty-list {
      padding: 20px;
      font-size: 2rem;
      color: #8c9094;
    }

    .todo-item {
      display: flex;
      justify-content: space-between;
      padding: 10px;
      border: 1px solid #a59b9b;
      margin-bottom: 10px;
      border-radius: 5px;
      border-color: #b3b3b3;

      &.completed {
        opacity: 0.5;
      }

      &:hover {
        background-color: #9bd0ff;
      }

      .name {
        font-size: 1.2rem;
      }

      .actions {
        button {
          margin-left: 10px;
        }
      }
    }
  }
}
