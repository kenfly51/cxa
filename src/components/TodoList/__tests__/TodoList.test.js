import React from "react";
import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { TodoList } from "../TodoList";
import Button from "react-bootstrap/Button";

Enzyme.configure({ adapter: new Adapter() });

describe("TodoList", () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation(init => [init, setState]);
  wrapper = Enzyme.shallow(<TodoList />);

  it("should render initial layout", () => {
    expect(wrapper.getElements()).toMatchSnapshot();
  });

  it("should add new todo when enter from textbox", () => {
    wrapper.find("#to-do-form").simulate("submit", {
      target: [{ value: "new todo 1" }],
      preventDefault: jest.fn()
    });

    expect(wrapper.find(".todo-item")).toHaveLength(1);
  });

  it("should add new todo when enter from textbox", () => {
    wrapper.find("#to-do-form").simulate("submit", {
      target: [{ value: "new todo 2" }],
      preventDefault: jest.fn()
    });

    expect(wrapper.find(".todo-item")).toHaveLength(2);
  });

  it("should delete todo when click delete button", () => {
    const deletebtn = wrapper.find(Button).first();
    deletebtn.simulate("click");

    expect(wrapper.find(".todo-item")).toHaveLength(1);
  });

  it("should mark todo as completed when click complete button", () => {
    const deletebtn = wrapper.find(Button).at(1);
    deletebtn.simulate("click");

    expect(wrapper.find(".todo-item").hasClass("completed")).toBe(true);
  });
});
