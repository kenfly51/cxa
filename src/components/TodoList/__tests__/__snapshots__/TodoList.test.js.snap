// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render initial layout 1`] = `
Array [
  <div
    className="todo-list"
  >
    <h3>
      To Do List
    </h3>
    <div
      className="cotainer"
    >
      <div
        className="search-bar"
      >
        <ForwardRef
          id="search-form"
          inline={false}
          onSubmit={[Function]}
        >
          <ForwardRef
            id="search-term"
            placeholder="search todo"
          />
        </ForwardRef>
      </div>
      <div
        className="empty-list"
      >
        I'm lonely now
      </div>
      <ForwardRef
        id="to-do-form"
        inline={false}
        onSubmit={[Function]}
      >
        <ForwardRef
          id="new-todo"
          placeholder="Enter new todo item..."
        />
      </ForwardRef>
    </div>
  </div>,
]
`;
